proposed/SerializerGoals5.0.md

```diff
@@ -0,0 +1,301 @@
+# Overview and goals
+
+April 21th, 2020.
+
+This document provides options and recommendations for 5.0 to handle overlapping serializer goals including reach, faster cold startup performance, reducing memory usage, and smaller size-on-disk.
+
+## Reach
+Achieved by limiting, removing or adding fallbacks for runtime code generation so that AOT optimized platforms including [Xamarin iOS](https://github.com/dotnet/runtime/issues/31326) can use System.Text.Json.
+
+In particular, iOS does not allow for any type of code generation whether Reflection.Emit, IL generation or IL->native through JITing. This means the IL is either compiled AOT (Ahead-Of-Time) to native code before deploying to the device, or that the IL is interpreted at runtime.
+
+The options to consider, in priority order include:
+1) Minimize the serializer's runtime code generation. For example, in 5.0, we are planning on avoiding `Type.MakeGenericType` when instantiating the converter for POCOs -- the converter will close the `JsonConverter<T>` generic at build time with `JsonConverter<System.Object>`.
+2) Add "loosely-typed" fallbacks to the serializer for cases that cannot be refactored; these fallacks will box to Object when using value types or use standard invoke-based reflection. This results in slower performance. Some serializers do have a Reflection.Emit fallback (Newtonsoft) and others do not (Jil and Utf8Json). System.Text.Json has a fallback for its usages of Reflection.Emit, but it does not have a fallback for its usage of `Type.MakeGenericType and .MakeGenericMethod` which does cause issues on iOS.
+3) Generate code for POCOs that avoids using the unsupported features. This also has higher runtime performance than the loosely-typed fallbacks.
+4) Add a high-performance replacement for targeted scenarios including property getters, property setters, constructors and as a stretch `Type.MakeGenericType and .MakeGenericMethod`.
+5) If possible, add or extend runtime interpreter support for missing cases. Xamarin (and .NET Native) has an interpreter mode.
+6) Throw PlatformNotSupportedException for when there is no alternative.
+
+## Faster startup performance
+
+This is important for scenarios including [micro-services](https://github.com/dotnet/runtime/issues/1568) where processes start and exit frequently. It is also important for benchmarks including TechEmpower.
+
+The serializer has fast startup compared to most other serializers including Newtonsoft's Json.NET, JIL and Utf8Json. Startup perf is mostly affected by generating IL - the Jil serializer is the most aggressive here and has the highest startup cost.
```

Comment:
```suggestion
The serializer has fast startup compared to most other serializers including Newtonsoft's Json.NET, JIL and Utf8Json. Startup perf is mostly affected by generating IL - the JIL serializer is the most aggressive here and has the highest startup cost.
```

Reply:
Actually, https://github.com/kevin-montrose/Jil uses `Jil` casing, so it is what we should use throughout the document as well, I think